src/main/java/evaluator/App.java

```diff
@@ -31,31 +34,80 @@
 
   @Parameters(index = "1", description = "Template file")
   private File templateFile;
+  
+  @Parameters(index = "2", description = "Folder for URL cache")
+  private String URLCacheFolderName;
+  
+  private boolean URLCacheEnabled = true;
+  private File URLCache;
 
   public static void main(String[] args) {
     CommandLine.call(new App(), args);
   }
-
+ 
   @Override
-  public Void call() throws Exception {
+  public Void call() throws Exception {    
+	try {
+	  URLCacheFolderName = URLCacheFolderName.trim();
+	  if (URLCacheFolderName.charAt(URLCacheFolderName.length()-1) != '/') {
+		  URLCacheFolderName = URLCacheFolderName.concat("/");
+	  }
+	  URLCache = new File(URLCacheFolderName);
+	  URLCache.mkdirs();
+	  URLCache.setWritable(true);
+	} catch (SecurityException e) {
+	  URLCacheEnabled = false;	
+      System.err.println("WARNING: Failed to create Writable URL Cache Folder " + URLCacheFolderName + ": " + e.getMessage() + ". Proceeding without Cache");  
+	}
+	    
     Template template = loadTemplate(templateFile);
     List<String> rulesNames = template.rules.stream().map(rule -> rule.name).collect(Collectors.toList());
     rulesNames.add(0, "url");
     System.out.println(String.join("\t", rulesNames));
     try (Stream<String> lines = Files.lines(urlFile.toPath())) {
       lines.forEachOrdered(line -> {
-        if(!line.trim().isEmpty()) {
+        if(!line.trim().isEmpty()) {        	
           try {
-            Document document = Jsoup.connect(line)
-                .userAgent("Mozilla/5.0 (compatible; Pinterestbot/1.0; +http://www.pinterest.com/bot.html)")
-                .get();
+        	Document document;
+        	
+        	if (URLCacheEnabled) {
+              Base32 base32 = new Base32();  
+        	  String encodedURL = new String(base32.encode(line.getBytes()));
```

Comment:
Extract a function? Maybe something like fetchDocument(url) that does the following:
Document document = null;
if(urlCacheEnabled) {
 document = readDocumentFromCache()
} 
if (document == null) {
// call Jsoup to fetch the HTML
...
  if(urlCacheEnabled) {
    writeDocumentToCache(url, document)
  }
}

Reply:
Yes, good idea. Done. Actually, I also extracted all the cache functionality into a new class, HTMLCache. I'll submit a new Pull request with the changes, because there are a lot more changes now.